client/components/sticky-panel/index.jsx

```diff
@@ -114,3 +87,80 @@ export default class extends React.Component {
 		);
 	}
 }
+
+class StickyPanelWithIntersectionObserver extends StickyPanelBase {
+	throttleOnResize = throttle( this.onWindowResize.bind( this ), RESIZE_RATE_IN_MS );
+
+	state = {
+		isSticky: false,
+		spacerHeight: 0,
+		blockWidth: 0,
+	};
+
+	onIntersection = afterLayoutFlush( entries => {
+		if ( ! entries || ! entries[ 0 ] ) {
+			return;
+		}
+		const { intersectionRatio, rootBounds, boundingClientRect } = entries[ 0 ];
+		const isSticky = intersectionRatio < 1 && rootBounds.bottom > boundingClientRect.bottom;
+		this.updateStickyState( isSticky );
+	} );
+
+	componentDidMount() {
+		window.addEventListener( 'resize', this.throttleOnResize );
+
+		this.deferredMount = defer( () => {
+			this.observer = new IntersectionObserver( this.onIntersection, {
+				threshold: [ 0, 1 ],
+				rootMargin: `-${ this.calculateOffset() }px 0px 0px 0px`,
+			} );
+			this.observer.observe( ReactDom.findDOMNode( this ) );
+		} );
+	}
+
+	componentWillUnmount() {
+		if ( this.observer ) {
+			this.observer.disconnect();
+		}
+		this.onIntersection.cancel();
+		clearTimeout( this.deferredMount );
+		window.removeEventListener( 'resize', this.throttleOnResize );
+	}
+}
+
+class StickyPanelWithScrollEvent extends StickyPanelBase {
```

Comment:
I only had a very quick glance at this PR so far, so I only noticed this one: using inheritance (extending from anything else than `React.Component`) is very rare in the React world. Could the sticky panel be refactored into component composition?

Reply:
Ah, thanks for the tip! No worries, I'll take a look at refactoring it tomorrow!